azure_li_services/units/user.py

```diff
@@ -81,40 +81,47 @@ def create_or_modify_user(user):
         raise AzureHostedUserConfigDataException(
             'username missing in config {0}'.format(user)
         )
-    options = []
+    user_options = []
     system_users = Users()
     user_exists = system_users.user_exists(user['username'])
     if 'group' in user:
-        if not system_users.group_exists(user['group']):
-            system_users.group_add(user['group'], [])
-        options += [
-            '-g', user['group']
+        group_setup = user['group']
+        if not system_users.group_exists(group_setup['name']):
+            group_options = []
+            if 'id' in group_setup:
+                group_options += [
+                    '-g', '{0}'.format(group_setup['id'])
+                ]
```

Comment:
This is the problem. If the group has a name, which it must have, as you correctly point out, we use the user supplied ID, if it exists, from the config file without checking if a group with this ID already exists. But groupadd will complain if a group-id already exists. 

The previously provided example

```
# groupadd -g 1000 foo
groupadd: GID '1000' already exists
```
shows this condition. The test system has a group with ID 1000 but no group named "foo". The way the code is implemented it would issue the command above if the following config snippet were part of the config:

`group: {'name': 'foo', 'id':1000}`

and the group 'foo' did not exist.

We have to check for both the existence of group-name and the group-id if an id is given in the config file, there is no way around it.

I am not certain what the behavior should be if we have conflicting group-ids, I would say the chance that the group-name exists is small. However, I would not make that assumption for the group-id, unless Microsoft will always use large group-id numbers, maybe 10000. In any event we have to check and then decide what to do if we cannot create the group.

If we cannot create the group the storage setup will fail. While the OS install at this point, without the group, would be usable. It is not usable for the purposes of setting up the application as the storage configuration is critically important. @jeffaco will have to provide input on the behavior that is desired if group creation fails.

Reply:
I got your concern, but why do you want us to check this ? if a group X is requested and it does not exist we try to create it, if that does not work because the supplied id for this group is used by another group the call will fail and we report that as an error condition with a good error message.

If I check before for id X if it is used by some other group, what else could I do as just fail with an error message ? I don't get why you want to check a condition which is checked by the groupadd call itself.

Reply:
But you could apply the same logic to the name check if the group with the name exists we just get an error. You could simply collect the group settings form config and unconditionally run 'groupadd'. If it errors we report accordingly. So the only condition would be If-config-has-group-setting.

Checking for the name and not the id is confusing as the assumption I made is reasonable, I think. The assumption is "we check the name to avoid an error". But I guess the check is for "do we really need to create the group. I think a comment to that effect would be good. Something like

# Do we really need to create the group?

Reply:
correct the check for the group name is done to make sure we create that group with that name if not present such that the later user creation can be assigned to that group. I'm sorry but I don't see why that logic is confusing.

> You could simply collect the group settings form config and unconditionally run 'groupadd'.

I have to at least check if there are group settings for the user ```if 'group' in user:``` I really don't know how it could be expressed more readable in code. Next I check if that group name exists, if not create it, and in the end assign the requested group to the user creation options. If there is a problem in the group or user creation; Fail with an exception and a good message.

This logic is not confusing and imho doesn't need a descriptive comment. I consider

* ```# do we really need to create the group```

the same as

* ```if not system_users.group_exists(group_setup['name']):```

it got requested in the group_setup, it does not exist, so what else other than create it is the logic consequence.

This reads like a debate between Dr. Mc-Coy and Mr. Spock :)

Thanks for approving, I'll go ahead since I wanted to get things done and update the image build. By the way you asked me to contact Fabian, which I did but he is on vacation. I'll send you a mail for further details